### ISSUE TITLE
Server doesn't fully recover after a PhantomJS crash

### ISSUE THREAD

**Issue by mbenford** [prerender/prerender#95]
Hi. First off, congrats on this amazing project! :+1: 

We've been evaluating Prerender as an alternative to replace our in-house solution for handling AJAX crawling (which is, by the way, similar to yours to a degree), and we've been experiencing some issues, mostly related to PhantomJS, as far as I can tell.

The first issue regards the number of PhantomJS workers. We've figured out that any number greater than 1 causes the server to hang after a small amount of requests. It seems that [other people](http://www.jonben.net/prerender-io-the-highs-and-lows/) have experienced that as well.

And the second, more critical issue is related to the PhantomJS restarting process. After tweaking the timeouts and setting the `PHANTOM_CLUSTER_NUM_WORKERS` to 1, we've managed to make the server pretty stable for some requests, but that stability disappears as soon as PhantonJS crashes or is killed by Prerender. The following graph illustrates that:

![prerender](https://cloud.githubusercontent.com/assets/446867/3250584/9bff339e-f1a7-11e3-91b0-6c80d2bd10e4.jpg)

The server seems to never fully recover after a PhantomJS crash. My best guess is that when a request arrives while PhantomJS is still starting up, bad things happen.

I'm willing to debug Prerender's source code in order to fix or at least understand better why the second issue is happening, and I'm looking for some guidance on where to begin. I appreciate any help.

For what it's worth, we're running Prerender on a dedicated Ubuntu machine and using JMeter to generate one request per second, which is how often Googlebot hits our servers on average.

**Comment by sylido** [comment]
I tried using the [PhantomJS 2.0.0 Travis CI build on Ubuntu 12.04](https://s3.amazonaws.com/travis-phantomjs/phantomjs-2.0.0-ubuntu-12.04.tar.bz2) and it works a bit better than the previous PhantomJS version, but still getting a lot of crashes while handling a lot of requests, granted they still might have some memory issues in this version. You just need to extract the binary from the file into the phantomjs binary folder where your prerender project is installed (i.e. copy the phantomjs file to prerender/node_modules/phantomjs/bin/ to replace the old one)

**Comment by ryanricard** [comment]
I am testing Prerender with Phantom2 on a 24 core machine with 100gb of RAM. I'm working on a solution to keeping a MongoDB ~1,000,000 page cache warm. With a single Prerender cluster of 24 Phantoms, I am processing 3-5 pages per second (~7 second page loads). Phantom is crashing randomly and I am seeing the following Node cluster error that seems to be resulting in orphaned Phantom processes:
```
events.js:85
      throw er; // Unhandled 'error' event
            ^
Error: channel closed
    at ChildProcess.target.send (child_process.js:414:26)
    at sendHelper (cluster.js:676:8)
    at RoundRobinHandle.handoff (cluster.js:198:3)
    at RoundRobinHandle.distribute (cluster.js:184:20)
```
Any idea what is causing this error? Should it be logged as a new issue?

Also, would I be better off load balancing between multiple Prerender instances with fewer Phantoms (e.g. 10 Prerenders with 2 Phantoms each)?

**Comment by thoop** [comment]
@ryanricard I haven't seen that error yet. It probably has something to do with how we "tree kill" the worker so that PhantomJS gets cleaned up too. We use tree kill there to kill the worker and PhantomJS because we were seeing orphaned PhantomJS processes but we haven't been seeing them since using tree kill.

I'd say log that as a new issue and I can look into it. We load balance between hundreds of instances with fewer Prerender/PhantomJS processes so maybe that's why we haven't run into this.

**Comment by ryanricard** [comment]
Thanks, @thoop. I will log it. If you don't mind me asking, what are you using for load balancing?

**Comment by thoop** [comment]
We're built out completely on Amazon Web Services so we are using their Elastic Load Balancer. Works great with hundreds of servers.

**Comment by ryanricard** [comment]
Good to know. Thanks again, @thoop. I will try to give back once I find opportunities.

**Comment by graywithanA** [comment]
I think that I'm experiencing the crashing/memory leak issue that has been discussed here but I wanted to confirm here that you guys agree.  After the prerender server had been running for about 18 hours and processing 5 requests the 5th request got a 504.  Here's the log from that point:

```
2015-10-15T12:08:31.674Z getting robots.txt
2015-10-15T12:08:31.988Z got 504 in 314ms for robots.txt
2015-10-15T12:08:32.159Z getting
phantom stderr: PhantomJS has crashed. Please read the crash reporting guide at https://github.com/ariya/phantomjs/wiki/Crash-Reporting and file a bug report at https://github.com/ariya/phantomjs/issues/new with the crash dump file attached: /tmp/52ef486e-85f9-c8db-0f1643f6-28bd2f62.dmp

2015-10-15T12:08:32.402Z phantom crashed, restarting...
2015-10-15T12:08:32.402Z starting phantom on port [12302]
2015-10-15T12:08:32.731Z started phantom
2015-10-15T12:08:42.673Z PhantomJS restarted in the middle of this request. Aborting...
2015-10-15T12:08:42.674Z got 504 in 10515ms for
2015-10-15T17:02:38.921Z getting robots.txt
2015-10-15T17:02:39.235Z got 504 in 314ms for robots.txt
2015-10-15T21:03:21.730Z got 200 in 3014ms for <myURL>
2015-10-15T21:03:21.917Z worker 3 died, restarting!
2015-10-15T21:03:22.225Z starting phantom on port [12309]
2015-10-15T21:03:22.255Z Server running on port 24892
2015-10-15T21:03:22.680Z started phantom
2015-10-15T21:32:01.990Z getting robots.txt
2015-10-15T21:32:02.300Z got 504 in 309ms for robots.txt
2015-10-15T21:32:02.411Z worker 8 died, restarting!
2015-10-15T21:32:02.576Z starting phantom on port [12310]
2015-10-15T21:32:02.584Z Server running on port 24892
2015-10-15T21:32:02.707Z started phantom
2015-10-15T22:08:41.307Z getting <myURL>
2015-10-15T22:08:44.638Z got 200 in 3330ms for <myURL>

...
```
This worker died/starting phantom pattern went on until the last one at 23:51 at which point WebFaction killed my processes for because I was way over my allotted memory usage.

Is this the same behavior discussed above that requires that I restart the prerender server occasionally?  Any suggestions as to how often?

I'm using [Forever](https://github.com/foreverjs/forever) to keep the server running if that is of relevance.

Also, I had reduced the number of iterations to 2 because I originally though that the default 10 was causing the memory issue.  My testing server is alloted 512mb of RAM, how many iterations should I allow?

**Comment by thoop** [comment]
PhantomJS uses a lot of memory so keeping it under 512mb could be hard. I see you are getting 504s but it looks like that is for robots.txt. You shouldn't be sending txt files to Prerender.

Reducing the number of iterations to 2 is also why your server is crashing a lot.

**Comment by graywithanA** [comment]
Can you suggest a minimum amount of memory, ideal number of iterations and potential server restart period?

**Comment by thoop** [comment]
Ideally, you would use 20-50 iterations to keep from killing PhantomJS too often and affecting other requests in the process. You shouldn't have to restart your servers more than daily and ideally restart them weekly. I would suggest using something that gives at least 4GB of memory to let you have enough room to use 20-50 iterations. And that would be for 1 worker.

I wish it was better than that but running a headless browser is pretty resource intensive.

**Comment by graywithanA** [comment]
Oh wow.  It sounds like there was no hope for my 512mb setup then.  Thanks
for the info.

**Comment by thoop** [comment]
A 512mb server could handle light load. Does your hosting provider just restart servers that go over the limit? Or what do they do in that case?

**Comment by graywithanA** [comment]
They kill all the processes, no restart.  For what it's worth, we've signed
up for your service for this particular project so this is no longer an
immediate concern.  This info is still useful for future projects though.

So for a 512mb server would it be best to limit the number of workers
instead of the number of iterations or both?

**Comment by mkukik** [comment]
I tried to build our solution based on this package https://www.npmjs.com/package/phantom-html-to-pdf but due to time constrains I have opted for your server that we host inhouse. I found "phantom-html-to-pdf" to be very stable (maybe due to different circumstances)... maybe you can take a look at their approach.

**Comment by thoop** [comment]
@mkukik phantom-html-to-pdf looks great. Are you trying to use that to serve prerendered pages or are you actually saving them as pdf? It looks like their PhantomJS servers only allow one request at a time where we allow many requests to be loaded in parallel. If you have too many requests loading in parallel, you could see performance problems.

**Comment by mkukik** [comment]
We have this in production to generate PDF... when I was testing, it looked pretty stable even with many requests at once (not dying phantomjs instances) and it served all those requests with correct data.
Truth is that service receives HTML as string opposite to having to download main html from given URL. And no javascript rendering... so it was just my 2 cents... sorry if I am wasting your time.

**Comment by thoop** [comment]
Not a waste of time at all! I love seeing how other projects do things and phantom-html-to-pdf looks like a great project.

I suspect it does help that you are just rendering a string and no javascript. A lot of PhantomJS issues are due to, like you said, http requests and the associated issues that come with trying to render javascript. It looks like they aren't doing much different from our project that would cause it to be more stable.

**Comment by mkukik** [comment]
just one more thing... I had this PDF package running and stable and when I did some refactoring of the code, nothing too fancy just exporting modules differently, moving stuff around it was still serving but became unstable under load... I guess combination of clustering, modules, instances could create some issues... if I get chance I will try to get back to it and turn on JS and other stuff to see if it still performs

**Comment by thoop** [comment]
Yeah it looks like phantom-html-to-pdf tries to start quite a few phantomjs servers on the same machine, so you might run into issues if you don't have a really beefy server. The same can happen with Prerender.

**Comment by montekaka** [comment]
this work well on my local machine, but it doesn't output the mp3 link im expecting to see.

any idea?

```ruby
require 'watir-webdriver'
require 'nokogiri'

desc "scape unmute watir"
task :unmute => :environment do
	args = %w{--ignore-ssl-errors=true}
	url = 'http://onunmute.com/replay/?fynLmB0sUO'
	browser = Watir::Browser.new(:phantomjs, :args => args)	
	browser.goto url
	sleep 3
	p browser.audio.source.src
	puts "finish parsing unmute"
end
```

**Comment by thoop** [comment]
@montekaka Did you post your comment in the wrong repo? That has nothing to do with Prerender.

**Comment by thoop** [comment]
This should be fixed with our latest version of Prerender. In v4.0.3 we've upgrade to a new phantomjs -> node bridge that seems more stable.
